Get build progress updates through Chromatic notify service

justin-thurman opened the pull request:
# Description

This PR implements support for getting build progress updates through the Chromatic notify service, rather than the existing polling implementation. The CLI will now attempt to open a websocket connection, wait for messages indicating that the build is complete, and then make a request to the index to get final build information. If the notify service errors or is otherwise unavailable, the existing index polling implementation is used as a fallback.

# Manual QA

Using a local Chromatic instance that implements pushing the necessary updates to the websocket service and sets `CHROMATIC_NOTIFY_SERVICE_URL` to `wss://ws.dev-chromatic.com:444` and `WEBSOCKET_SERVICE_URL` (this env var to eventually change to `NOTIFY_SERVICE_URL`) to `https://ws.dev-chromatic.com:444`, I tested a few scenarios.

A basic story with a single test, showing a simple happy path scenario. Debug logs shown:
![image](https://github.com/user-attachments/assets/b5f6f207-83fb-46a0-b9e7-8fcbd2a10387)

A larger build with multiple stories, showing the UI state updates working as expected. A few example screenshots, including the resulting debug logs:
![image](https://github.com/user-attachments/assets/0e507644-16ca-4e09-9e8c-abe33a43ed0a)

![image](https://github.com/user-attachments/assets/4369c357-01ec-436f-aace-2d1b3ccbfcd1)

![image](https://github.com/user-attachments/assets/f58e93af-42e7-4d49-aa01-f89e9fe188f2)

A build ran without the notify service running, showing error logging and that the existing polling implementation is used as a fallback:
![image](https://github.com/user-attachments/assets/8a368bf0-a166-4aa9-895c-203959ad2d5e)

A build where I added `foo` to the subscribe URL, which results in subscribing to a channel that never receives any messages. This leads to a `408 Response Timeout`, which you can see handling for:
![image](https://github.com/user-attachments/assets/1f8723b5-dcc4-4a2c-9e8a-6bd29b2a9db1)

And a build where I passed an invalid auth header, showing the 403 handling:
<img width="804" alt="image" src="https://github.com/user-attachments/assets/dd0aa388-9bd7-42c6-81fb-2511d5f4bd9d" />

And finally, a full build of the monorepo on staging:
![image](https://github.com/user-attachments/assets/c423ebf8-8b5c-42c9-a04f-8c3729fb6664)

...

![image](https://github.com/user-attachments/assets/6e6b346f-f2ed-493a-9229-4777fff3f0e6)

<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>13.1.0--canary.1191.16031610743.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install chromatic@13.1.0--canary.1191.16031610743.0
  # or 
  yarn add chromatic@13.1.0--canary.1191.16031610743.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

justin-thurman commented on node-src/lib/getEnvironment.ts:
```diff
@@ -24,6 +25,7 @@ const {
   CHROMATIC_DNS_SERVERS = '',
   CHROMATIC_HASH_CONCURRENCY = '48',
   CHROMATIC_INDEX_URL = 'https://index.chromatic.com',
+  CHROMATIC_NOTIFY_SERVICE_URL = 'wss://ws.chromatic.com:444',
```
I'm guessing this will be the prod endpoint. If not, we'll have to update this.

justin-thurman commented on node-src/lib/getEnvironment.ts:
```diff
@@ -24,6 +25,7 @@ const {
   CHROMATIC_DNS_SERVERS = '',
   CHROMATIC_HASH_CONCURRENCY = '48',
   CHROMATIC_INDEX_URL = 'https://index.chromatic.com',
+  CHROMATIC_WEBSOCKET_SERVICE_URL = 'wss://ws.chromatic.com:444',
```
I'm guessing this will be the prod endpoint. If not, we'll have to update this.

codykaup replied:
I want to say this is still open for discussion. The original setup was using `notify.chromatic.com` but we can make it whatever makes sense. WDYT?

justin-thurman replied:
At this point, I think we should call it websocket instead of notify, as that's how we're referring to it in most of our current code, repo names, documentation, etc. But if we want to switch back to notify, that's fine with me.

Though this does get me thinking about the current implementation and whether we should merge it. Right now, anyone upgrading to this version would get error logs, since we throw and then error log when we fail to handshake the websocket server. So I think we could do one of two things here: 1) not merge this until the websocket server is live in production, or 2) handle websocket closure code 1006 with a debug log instead of an error log. This is the code we get when we fail to connect, though theoretically, it could be raised in other scenarios too, afaik.

As I'm talking through it, I think we should handle closure 1006 differently regardless. I think we should have a log message that gives a bit more detail: something like `failed to connect, falling back to polling`. But it's an open question whether this should be an error log or a debug log. Thoughts?

codykaup commented on node-src/lib/waitForBuildToComplete.ts:
```diff
@@ -0,0 +1,105 @@
+  });
+  return await new Promise((resolve, reject) => {
+    subscriber.on('open', () => {
+      log.debug(`websocket handshake successful at ${url}`);
```
Why are debug logs showing up in a normal run (not using `--debug`)

justin-thurman replied:
Oh, yeah, in the WIP version I first pushed up, everything was `info`.

justin-thurman commented:
Converting this back to draft. I'm going to wait for the index side to be merged before reopening this for review, just in case I need to change things on my end. Don't want to ask for re-reviews all the time. cc @codykaup

justin-thurman commented on node-src/tasks/snapshot.ts:
```diff
@@ -97,10 +106,40 @@ export const takeSnapshots = async (ctx: Context, task: Task) => {
+    if (error instanceof NotifyConnectionError) {
+      ctx.log.error('Failed to connect to notify service, falling back to polling');
+    } else if (error instanceof NotifyServiceMessageTimeoutError) {
+      ctx.log.error('Timed out waiting for message from notify service, falling back to polling');
+      Sentry.captureException(error);
```
Not sure if we want to capture all exceptions, or if that would be too noisy. 🤔

justin-thurman commented:
@codykaup, this is ready for re-review, though I still don't plan to merge it until the notify service is deployed in prod and staging, so if you'd like to hold off on reviewing until then, feel free. I redid all the manual testing and included updated instructions and screenshots in the PR description. I also added handling of the `408 Request Timeout` from the notify service. I split that logic in a separate commit, if you'd like to review them separately.

codykaup commented:
> @codykaup, this is ready for re-review, though I still don't plan to merge it until the notify service is deployed in prod and staging, so if you'd like to hold off on reviewing until then, feel free. I redid all the manual testing and included updated instructions and screenshots in the PR description. I also added handling of the `408 Request Timeout` from the notify service. I split that logic in a separate commit, if you'd like to review them separately.

Awesome, thank you! I'll take a peek after I port the server configs over so I can unblock the staging/prod setup.

justin-thurman commented:
@codykaup, I updated the logging to `debug`, as we discussed, and ran a full build of the monorepo on staging. All working as expected. Added screenshots to the PR description.

justin-thurman commented on node-src/tasks/snapshot.ts:
```diff
@@ -97,10 +107,49 @@ export const takeSnapshots = async (ctx: Context, task: Task) => {
+    }
+  };
+  try {
+    await waitForBuildToComplete({
```
4ed180299c08ed01b33baca480f4ca900990972d

justin-thurman commented on node-src/lib/waitForBuildToComplete.ts:
```diff
@@ -0,0 +1,214 @@
+      log.debug(`notify service handshake successful at ${url}`);
+    });
+
+    subscriber.on('unexpected-response', (_request, response) => {
```
The types package is already present, but I did update the type hints in a few places to hopefully make things more clear. a8d2795ec9b34835038d790aa3499556f6603080

justin-thurman commented:
> This feels more like a minor bump than a patch. WDYT?

Oh, for sure. I clicked the wrong one. Meant `minor`.
